File: lib/vfs/applesingle.cc
#include "lib/core/globals.h"
#include "lib/core/bytes.h"
#include "lib/vfs/applesingle.h"

static constexpr uint32_t APPLESINGLE_MAGIC = 0x00051600;
static constexpr uint32_t APPLESINGLE_VERSION = 0x00020000;

void AppleSingle::parse(const Bytes& bytes)
{
    ByteReader br(bytes);
    if (br.read_be32() != APPLESINGLE_MAGIC)
        throw InvalidFileException();
    if (br.read_be32() > APPLESINGLE_VERSION)
        throw InvalidFileException();

    br.skip(16);
    int entries = br.read_be16();
    for (int i = 0; i < entries; i++)
    {
        uint32_t entryType = br.read_be32();
        uint32_t offset = br.read_be32();
        uint32_t length = br.read_be32();

        switch (entryType)
        {
            case 1:
                data = bytes.slice(offset, length);
                break;

            case 2:
                rsrc = bytes.slice(offset, length);
                break;

            case 9:
            {
                Bytes finderinfo = bytes.slice(offset, length);
                type = finderinfo.slice(0, 4);
                creator = finderinfo.slice(4, 4);
                break;
            }

            default:
                break;
        }
    }
}

Bytes AppleSingle::render()
{
    Bytes result;
    ByteWriter bw(result);

    bw.write_be32(APPLESINGLE_MAGIC);
    bw.write_be32(APPLESINGLE_VERSION);
    bw.append(Bytes(16)); /* padding */
    bw.write_be16(3);     /* number of entries */

    bw.write_be32(9);    /* finder info */
    bw.write_be32(0x3e); /* finder info offset */
    bw.write_be32(32);   /* finder info length */

    bw.write_be32(1);           /* data fork */
    bw.write_be32(OVERHEAD);    /* data fork offset */
    bw.write_be32(data.size()); /* data fork length */

    bw.write_be32(2);                      /* resource fork */
    bw.write_be32(OVERHEAD + data.size()); /* resource fork offset */
    bw.write_be32(rsrc.size());            /* resource fork length */

    bw.append(type.slice(0, 4));
    bw.append(creator.slice(0, 4));
    bw.append(Bytes(32 - 8));

    bw.append(data);
    bw.append(rsrc);

    return result;
}

File: java/com/cowlark/fluxengine/vfs/AppleSingle.java
package com.cowlark.fluxengine.vfs;

import com.cowlark.fluxengine.core.ByteReader;
import com.cowlark.fluxengine.core.ByteWriter;
import com.cowlark.fluxengine.core.Bytes;

/**
 * AppleSingle container, ported from {@code old_cpp_version/lib/vfs/applesingle.cc}.
 * Holds a data fork, resource fork, and Finder metadata (type/creator).
 */
public class AppleSingle
{
    public static final int OVERHEAD = 0x5e;

    private static final int APPLESINGLE_MAGIC = 0x00051600;
    private static final int APPLESINGLE_VERSION = 0x00020000;
    public Bytes data = new Bytes();
    public Bytes rsrc = new Bytes();
    public Bytes type = new Bytes();
    public Bytes creator = new Bytes();

    public void parse(Bytes bytes)
    {
        ByteReader br = new ByteReader(bytes);
        if (br.readBe32() != APPLESINGLE_MAGIC)
            throw new InvalidFileException("bad AppleSingle magic");
        if (Integer.compareUnsigned(br.readBe32(), APPLESINGLE_VERSION) > 0)
            throw new InvalidFileException("unsupported AppleSingle version");

        br.skip(16);
        int entries = br.readBe16();
        for (int i = 0; i < entries; i++)
        {
            int entryType = br.readBe32();
            int offset = br.readBe32();
            int length = br.readBe32();

            switch (entryType)
            {
                case 1:
                    data = bytes.slice(offset, length);
                    break;

                case 2:
                    rsrc = bytes.slice(offset, length);
                    break;

                case 9:
                {
                    Bytes finderInfo = bytes.slice(offset, length);
                    type = finderInfo.slice(0, 4);
                    creator = finderInfo.slice(4, 4);
                    break;
                }

                default:
                    break;
            }
        }
    }

    public Bytes render()
    {
        Bytes result = new Bytes();
        ByteWriter bw = new ByteWriter(result);

        bw.writeBe32(APPLESINGLE_MAGIC);
        bw.writeBe32(APPLESINGLE_VERSION);
        bw.pad(16);
        bw.writeBe16(3);

        bw.writeBe32(9);
        bw.writeBe32(0x3e);
        bw.writeBe32(32);

        bw.writeBe32(1);
        bw.writeBe32(OVERHEAD);
        bw.writeBe32(data.size());

        bw.writeBe32(2);
        bw.writeBe32(OVERHEAD + data.size());
        bw.writeBe32(rsrc.size());

        bw.write(type.slice(0, 4));
        bw.write(creator.slice(0, 4));
        bw.pad(32 - 8);

        bw.write(data);
        bw.write(rsrc);

        return result;
    }

    public static class InvalidFileException extends IllegalArgumentException
    {
        public InvalidFileException()
        {
            super("invalid AppleSingle file");
        }

        public InvalidFileException(String message)
        {
            super(message);
        }
    }
}
